nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-shell-authorizer/src/main/java/org/apache/nifi/authorization/ShellUserGroupProvider.java

```diff
@@ -0,0 +1,367 @@
+                }
+            });
+
+        synchronized (usersById) {
+            usersById.clear();
+            usersById.putAll(byId);
+        }
+
+        synchronized (usersByName) {
+            usersByName.clear();
+            usersByName.putAll(byName);
+            logger.debug("refreshUsers users now size: " + usersByName.size());
+        }
+    }
+
+    private void refreshGroups() {
+        Map<String, Group> groups = new HashMap<>();
+        List<String> lines;
+
+        try {
+            lines = ShellRunner.runShell(selectedShellCommands.getGroupsList());
+        } catch (final IOException ioexc) {
+            logger.error("refreshGroups list groups shell exception: " + ioexc);
+            return;
+        }
+
+        lines.forEach(line -> {
+                String[] record = line.split(":");
+                if (record.length > 1) {
+                    Set<String> users = new HashSet<>();
+                    String name = record[0], id = record[1];
+
+                    try {
+                        List<String> userLines = ShellRunner.runShell(String.format(selectedShellCommands.getGroupMembers(), name));
+                        if (userLines.size() > 0) {
+                            users.addAll(Arrays.asList(userLines.get(0).split(",")));
+                        }
+                    } catch (final IOException ioexc) {
+                        logger.error("refreshGroups list membership shell exception: " + ioexc);
+                    }
+
+                    if (name != null && id != null && !name.equals("") && !id.equals("")) {
+                        Group group = new Group.Builder().name(name).identifier(id).addUsers(users).build();
+                        groups.put(id, group);
+                        logger.debug("refreshed group: " + group);
+                    } else {
+                        logger.warn("null or empty group name: " + name + " or id: " + id);
+                    }
+                }
+            });
+
+        List<String> sublines;
+        try {
+            sublines = ShellRunner.runShell(selectedShellCommands.getUsersList());
+        } catch (final IOException ioexc) {
+            logger.error("refreshGroups list groups shell exception: " + ioexc);
+            return;
+        }
+
+        sublines.forEach(line -> {
+                String[] subrecord = line.split(":");
+
+                if (subrecord.length > 2) {
+                    String primaryGid = subrecord[2];
+                    Group primaryGroup = groups.get(primaryGid);
+                    Group group;
+
+                    if (primaryGroup == null) {
+                        logger.warn("user: " + subrecord[0] + " primary group not found");
+                    } else {
+                        Set<String> groupUsers = primaryGroup.getUsers();
+                        if (!groupUsers.contains(subrecord[0])) {
+                            Set<String> secondSet = new HashSet<>();
```

Comment:
I was not following this until I realized (I think) that a group's users may not include a user who has that group specified via the _primary group ID_ value in the user account. In that case, the logic is to add the current user (`subrecord[0]`) to the group's user list before building the `Group` object. 

Is there a reason a new `Set<String>` is constructed instead of just adding the current user to `groupUsers`?

Reply:
I think `groupUsers` is read-only; I get exceptions when I try to add a value to it.